src/atomWithSearchParams.ts

```diff
@@ -0,0 +1,99 @@
+import type { SetStateAction, WritableAtom } from 'jotai/vanilla';
+import { atom } from 'jotai/vanilla';
+import { atomWithLocation } from './atomWithLocation';
+
+// Create an atom for managing location state, including search parameters.
+const locationAtom = atomWithLocation();
+
+/**
+ * Creates an atom that manages a single search parameter.
+ *
+ * The atom automatically infers the type of the search parameter based on the
+ * type of `defaultValue`.
+ *
+ * The atom's read function returns the current value of the search parameter.
+ * The atom's write function updates the search parameter in the URL.
+ *
+ * @param key - The key of the search parameter.
+ * @param defaultValue - The default value of the search parameter.
+ * @returns A writable atom that manages the search parameter.
+ */
+export const atomWithSearchParams = <T>(
```

Comment:
Alright, then limit only to support string, number and boolean.
```suggestion
export const atomWithSearchParams = <T extends string | number | boolean>(
```
and simply the code please?

Reply:
I've tried, but I can't figure out why this strange type is being returned in TypeScript. I haven't yet pinpointed the reason.

Atom:
`const pageAtom = atomWithSearchParams("page", 1)`

Expected:

`const pageAtom: WritableAtom<number, [SetStateAction<number>], void>`

Got:

`const pageAtom: WritableAtom<1, [SetStateAction<1>], void>`

Reply:
I think in this case you need to do `atomWithSearchParams<number>("page", 1)` or `atomWithSearchParams("page", 1 as number)`. Do you have any other issues with this change?

Reply:
Actually, union type doesn't work. We should overload the function.